In [1]:
#importing necessary libraries
import pandas as pd
import numpy as np

### Q1: What's the version of Pandas that you installed?

In [2]:
pd.__version__

'2.2.3'

In [3]:
# Loading the dataset
Laptops= pd.read_csv('Laptops Price.csv')
Laptops.head(10)

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   
5  MSI Crosshair 17 C12VF-264XES Intel Core i7-12...    New     MSI   
6  Lenovo Thinkpad E14 Gen 4 AMD Ryzen 5 5625U/8G...    New  Lenovo   
7  ASUS VivoBook 15 F515JA-EJ2882W Intel Core i7-...    New    Asus   
8  Medion Akoya E15415 Intel Core i5-10210U/8GB/2...    New  Medion   
9  HP Victus 16-d1038ns Intel Core i7-12700H/16GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   
5   Crosshair  Intel Core i7   32     1000          SSD  RTX 4060    17.3   
6    ThinkPad    AMD Ryzen 5    8      256          SSD       NaN    14.0   
7    VivoBook  Intel Core i7    8      512          SSD       NaN    15.6   
8       Akoya  Intel Core i5    8      256          SSD       NaN    15.6   
9      Victus  Intel Core i7   16      512          SSD  RTX 3050    16.1   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01  
5    No      1699.00  
6    No       909.00  
7    No       809.01  
8    No       519.00  
9    No      1149.00

### Q2. Records count 
#### How many records are in the dataset?

In [4]:
#Counting the number of records
Laptops.shape[0]

2160

Columns are called Fields in a dataset while rows are called Records. The total number of record count is equal to the total number of Rows in the dataset hence, there are 2160 records in the dataset

### Q3. Laptop brands

#### How many laptop brands are presented in the dataset?

In [5]:
Laptops['Brand'].value_counts()

Brand
Asus                415
HP                  368
Lenovo              366
MSI                 308
Acer                137
Apple               116
Dell                 84
Microsoft            77
Gigabyte             48
Razer                37
Medion               32
LG                   32
Alurin               29
PcCom                24
Samsung              22
Dynabook Toshiba     19
Vant                 11
Primux                8
Deep Gaming           8
Innjoo                6
Thomson               4
Prixton               3
Millenium             2
Denver                1
Jetwing               1
Realme                1
Toshiba               1
Name: count, dtype: int64

In [6]:
#Counting the number of Laptop Brands
Laptops['Brand'].nunique()

27

There is a total of 27 `Laptop brands` in the dataset

### Q4. Missing values
#### How many columns in the dataset have missing values?

In [7]:
(Laptops.isnull().sum() !=0).sum()

3

In [8]:
Laptops.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

Three columns have missing values. `Storage type`, `GPU` and `Screen` columns.

### Q5. Maximum final price
#### What's the maximum final price of Dell notebooks in the dataset?

In [9]:
Laptops[Laptops['Brand']== 'Dell'].head()

Laptop Status Brand     Model  \
161  Dell Vostro 3510 Intel Core i3-1115G4/8 GB/256...    New  Dell    Vostro   
271  Dell Vostro 3510 Intel Core i5-1135G7/8 GB/256...    New  Dell    Vostro   
281  Dell Vostro 3520 Intel Core i7-1255U/16GB/512G...    New  Dell    Vostro   
319  Dell Latitude 3520 Intel Core i5-1135G7/8GB/25...    New  Dell  Latitude   
328  Dell Vostro 3510 Intel Core i5-1135G7/16 GB/51...    New  Dell    Vostro   

               CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
161  Intel Core i3    8      256          SSD  NaN    15.6    No       599.00  
271  Intel Core i5    8      256          SSD  NaN    15.6    No       690.15  
281  Intel Core i7   16      512          SSD  NaN    15.6    No       889.00  
319  Intel Core i5    8      256          SSD  NaN    15.6    No       789.00  
328  Intel Core i5   16      512          SSD  NaN    15.6    No       869.00

In [10]:
Laptops[Laptops['Brand'] == 'Dell']['Final Price'].max()

3936.0

The maximum final Price of `Dell` notebooks is 3936

### Q6. Median value of Screen

1. Find the median value of `Screen` column in the dataset.
2. Next, calculate the most frequent value of the same `Screen` column.
3. Use `fillna` method to fill the missing values in `Screen` column with the most frequent value from the previous step.
4. Now, calculate the median value of `Screen` once again.

Has it changed?

> Hint: refer to existing `mode` and `median` functions to complete the task.

- Yes
- No

In [11]:
Laptops['Screen'].value_counts()

Screen
15.60    1009
14.00     392
16.00     174
17.30     161
13.30     131
16.10      48
17.00      33
13.00      27
15.00      21
13.50      19
13.40      19
11.60      16
14.20      14
12.30      13
14.10      11
13.60      11
16.20      10
15.30       8
10.50       7
14.40       6
12.40       6
15.40       5
12.00       4
18.00       3
14.50       3
13.90       2
12.50       1
10.10       1
10.95       1
Name: count, dtype: int64

In [12]:
#median value
median_screen= Laptops['Screen'].median()
median_screen

15.6

In [13]:
#most frequent value
modal_screen= Laptops['Screen'].mode()[0]
modal_screen

15.6

In [14]:
Laptops['Screen'].fillna(modal_screen).median()

15.6

No, the median remains unchanged

### Q7. Sum of weights

1. Select all the "Innjoo" laptops from the dataset.
2. Select only columns `RAM`, `Storage`, `Screen`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the sum of all the elements of the result?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- 0.43
- 45.29
- 45.58
- 91.30

In [15]:
Innjoo_df= Laptops[Laptops['Brand']== 'Innjoo'].head()
Innjoo_df= Innjoo_df[['RAM','Storage','Screen']]
Innjoo_df

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1

In [16]:
X= Innjoo_df.to_numpy()
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1]])

In [17]:
XTX= X.T @ X
XTX

array([[2.16000e+02, 7.55200e+03, 4.75200e+02],
       [7.55200e+03, 3.52256e+05, 1.55904e+04],
       [4.75200e+02, 1.55904e+04, 1.08315e+03]])

In [18]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.81442446e-01, -1.56768809e-03, -1.00909907e-01],
       [-1.56768809e-03,  1.65537172e-05,  4.49509585e-04],
       [-1.00909907e-01,  4.49509585e-04,  3.87244183e-02]])

In [19]:
y= np.array([1100, 1300, 800, 900, 1000, 1100])
y

array([1100, 1300,  800,  900, 1000, 1100])